// @wagmi/vue/dist/esm/composables/useSwitchConnection.js
import { useMutation } from '@tanstack/vue-query';
import { switchConnectionMutationOptions, } from '@wagmi/core/query';
import { computed } from 'vue';
import { useConfig } from './useConfig.js';
import { useConnections } from './useConnections.js';
/** https://wagmi.sh/vue/api/composables/useSwitchConnection */
export function useSwitchConnection(parameters = {}) {
    const config = useConfig(parameters);
    const connections = useConnections({ config });
    const mutationOptions = switchConnectionMutationOptions(config);
    const mutation = useMutation({ ...parameters.mutation, ...mutationOptions });
    return {
        ...mutation,
        connectors: computed(() => connections.value.map((connection) => connection.connector)),
        mutate: mutation.mutate,
        mutateAsync: mutation.mutateAsync,
        switchAccount: mutation.mutate,
        switchAccountAsync: mutation.mutateAsync,
        switchConnection: mutation.mutate,
        switchConnectionAsync: mutation.mutateAsync,
    };
}
