// @wagmi/vue/dist/esm/exports/index.js
////////////////////////////////////////////////////////////////////////////////
// Plugin
////////////////////////////////////////////////////////////////////////////////
// biome-ignore lint/performance/noBarrelFile: entrypoint module
export { configKey, WagmiPlugin } from '../plugin.js';
////////////////////////////////////////////////////////////////////////////////
// Errors
////////////////////////////////////////////////////////////////////////////////
export { BaseError } from '../errors/base.js';
export { WagmiInjectionContextError, WagmiPluginNotFoundError, } from '../errors/plugin.js';
////////////////////////////////////////////////////////////////////////////////
// Composables
////////////////////////////////////////////////////////////////////////////////
export { useBalance, } from '../composables/useBalance.js';
export { useBlockNumber, } from '../composables/useBlockNumber.js';
export { useBytecode, } from '../composables/useBytecode.js';
export { useChainId, } from '../composables/useChainId.js';
export { useChains, } from '../composables/useChains.js';
export { useClient, } from '../composables/useClient.js';
export { useConfig, } from '../composables/useConfig.js';
export { useConnect, } from '../composables/useConnect.js';
export { 
/** @deprecated use `useConnection` instead */
useConnection as useAccount, useConnection, } from '../composables/useConnection.js';
export { 
/** @deprecated use `useConnectionEffect` instead */
useConnectionEffect as useAccountEffect, useConnectionEffect, } from '../composables/useConnectionEffect.js';
export { useConnections, } from '../composables/useConnections.js';
export { useConnectorClient, } from '../composables/useConnectorClient.js';
export { useConnectors, } from '../composables/useConnectors.js';
export { useDisconnect, } from '../composables/useDisconnect.js';
export { useEnsAddress, } from '../composables/useEnsAddress.js';
export { useEnsAvatar, } from '../composables/useEnsAvatar.js';
export { useEnsName, } from '../composables/useEnsName.js';
export { useEstimateGas, } from '../composables/useEstimateGas.js';
export { useReadContract, } from '../composables/useReadContract.js';
export { useReconnect, } from '../composables/useReconnect.js';
export { useSendTransaction, } from '../composables/useSendTransaction.js';
export { useSignMessage, } from '../composables/useSignMessage.js';
export { useSignTypedData, } from '../composables/useSignTypedData.js';
export { useSimulateContract, } from '../composables/useSimulateContract.js';
export { useSwitchChain, } from '../composables/useSwitchChain.js';
export { 
/** @deprecated use `useSwitchConnection` instead */
useSwitchConnection as useSwitchAccount, useSwitchConnection, } from '../composables/useSwitchConnection.js';
export { useTransaction, } from '../composables/useTransaction.js';
export { useTransactionReceipt, } from '../composables/useTransactionReceipt.js';
export { useWaitForTransactionReceipt, } from '../composables/useWaitForTransactionReceipt.js';
export { useWatchBlockNumber, } from '../composables/useWatchBlockNumber.js';
export { useWatchContractEvent, } from '../composables/useWatchContractEvent.js';
export { useWriteContract, } from '../composables/useWriteContract.js';
////////////////////////////////////////////////////////////////////////////////
// @wagmi/core
////////////////////////////////////////////////////////////////////////////////
export { ChainNotConfiguredError, ConnectorAccountNotFoundError, ConnectorAlreadyConnectedError, ConnectorChainMismatchError, ConnectorNotFoundError, ConnectorUnavailableReconnectingError, 
// Utilities
cookieStorage, cookieToInitialState, createConfig, createConnector, createStorage, 
// Transports
custom, deepEqual, deserialize, fallback, http, injected, mock, noopStorage, ProviderNotFoundError, parseCookie, SwitchChainNotSupportedError, serialize, unstable_connector, webSocket, } from '@wagmi/core';
////////////////////////////////////////////////////////////////////////////////
// Version
////////////////////////////////////////////////////////////////////////////////
export { version } from '../version.js';
